src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -266,6 +264,96 @@ private void HandleChannelException(IConnection connection, Exception ex)
             }
         }
 
+        private bool IsNotMaster(ServerErrorCode code, string message)
+        {
+            switch (code)
+            {
+                case ServerErrorCode.NotMaster: // 10107
+                case ServerErrorCode.NotMasterNoSlaveOk: // 13435
+                    return true;
+            }
+
+            if (message != null)
+            {
+                if (message.IndexOf("not master", StringComparison.OrdinalIgnoreCase) != -1 &&
+                    message.IndexOf("not master or secondary", StringComparison.OrdinalIgnoreCase) == -1)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool IsNotMasterOrRecovering(ServerErrorCode code, string message)
+        {
+            return IsNotMaster(code, message) || IsRecovering(code, message);
+        }
+
+        private bool IsRecovering(ServerErrorCode code, string message)
+        {
+            switch (code)
+            {
+                case ServerErrorCode.InterruptedAtShutdown: // 11600
+                case ServerErrorCode.InterruptedDueToReplStateChange: // 11602
+                case ServerErrorCode.NotMasterOrSecondary: // 13436
+                case ServerErrorCode.PrimarySteppedDown: // 189
+                case ServerErrorCode.ShutdownInProgress: // 91
+                    return true;
+            }
+
+            if (message != null)
+            {
+                if (message.IndexOf("not master or secondary", StringComparison.OrdinalIgnoreCase) != -1 ||
+                    message.IndexOf("node is recovering", StringComparison.OrdinalIgnoreCase) != -1)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool ShouldInvalidateServer(Exception exception)
+        {
+            if (__invalidatingExceptions.Contains(exception.GetType()))
+            {
+                return true;
```

Comment:
If we return true here we're doing exactly the same thing as before.

But if we don't return we're going to check for some additional scenarios as per the SDAM spec.